### PR TITLE
Tree cleanup

### PR DESCRIPTION
**Pull request by mlyle**
* Work towards #231: factor UAVOs common
* Remove 'pathfollower' module
* Remove faultsettings from CC/Naze32.  I kinda like the fault injection module, but it probably belongs on larger targets today.  Either that or we should totally remove it.  (it's currently compiled against nothing)

### CODE REVIEW

**Comment by peabody124** [comment]
Love the LOC count :)

**Comment by BrainFPV** [comment]
LGTM :)

**Comment by mlyle** [comment]
> I would like to figure out how to get rid of UAVO_NAV (and similar Makefile flags). In an ideal world, including the appropriate navigation module in the Makefile would be the only requirement necessary to build a working binary.

I think the way to do this is to have module specific makefiles that enabling a module includes.  Then they can either set flags or declare dependencies on UAVOs in there.  They can also declare anything about how they want their code compiled.

> Pardon my ignorance, but why aren't all UAVOs included for all firmware targets? Shouldn't the linker handle stripping out the unused functions, thus removing the need for per-target specification?

So we need to turn them off or on somewhere so that they don't use memory on smaller targets.  So that's why we have a UAVOBJ_Init_... flag.

And as long as we're doing that, we might as well not build the corresponding source files and save a little build time.  The makefile produces both lists (the list of sourcefiles and the list of _Init_ flags) adjacent to each other.

One alternative option would be to compile all the uavos (except uavobjectsinit) into a library before building the flight targets.  The different ARM targets don't really need different compile flags for the uavos-- there's not a lot of prospect for improvement from architecture-specific code. The simulator adds a small bit of complexity here.  I'm not sure it's worth it.

Basically at this point my objective is to just deduplicate and factor out stuff.  There's a lot of things that can be done to potentially clean stuff up, but the duplication is worst (and makes it harder to do those other things).

**Comment by mlyle** [comment]
Also I would like to end run-time registration of UAVOs.  We have a lot of code to build up a datastructure and figure what what uavo is from an id, etc.  We can avoid all that just by generating switch statements at compile time and run this directly in code.

**Comment by kubark42** [comment]
> So we need to turn them off or on somewhere so that they don't use memory on smaller targets.

Grumble, grumble. There's a can of worms here about obsolete targets interfering with future progress, but that's for another day.

> One alternative option would be to compile all the uavos (except uavobjectsinit) into a library before building the flight targets. 

Compiling all UAVOs with ```-fdata-sections -ffunction-sections````and linking the firmware blob with ```-Wl,--gc-sections``` in theory strips out the unused functions. That *should* get the unused UAVObjects, right?

> Also I would like to end run-time registration of UAVOs. We have a lot of code to build up a datastructure and figure what what uavo is from an id, etc. We can avoid all that just by generating switch statements at compile time and run this directly in code.

Good idea!

**Comment by mlyle** [comment]
> Grumble, grumble. There's a can of worms here about obsolete targets interfering with future progress, but that's for another day.

I like CC3D/Naze32.  It's nice to have a cheap widely available target.  Also it's debatable whether we should have hw objects for unsupported board types, etc.

> Compiling all UAVOs with -fdata-sections -ffunction-sections and linking the firmware blob with -Wl,--gc-sections in theory strips out the unused functions. That should get the unused UAVObjects, right?

Yes, but if they're registered they're "used".

There's a "clever" way to do the initialization/registration dynamically using linker scripts, but I don't like being too "clever."

**Comment by kubark42** [comment]
> Yes, but if they're registered they're "used".

Ah, I see. So the autogenerated ```uavobjectsinit.c``` is the culprit here. And the only way to formally keep the UAVObjects out is not include them in the first place. Sigh.

P.S. I kind of doubt it, but I wonder if there's a function to tell ```—gc-sections``` to ignore certain files when deciding if something is used or not.

**Comment by mlyle** [comment]
> P.S. I kind of doubt it, but I wonder if there's a function to tell —gc-sections to ignore certain files when deciding if something is used or not.

If there was, the call in the uavobjectsinit.c would fail to resolve to a symbol.  The linker is not anyone's friend.
